src/main/java/net/snowflake/client/core/QueryContextCache.java

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2012-2022 Snowflake Computing Inc. All rights reserved.
+ */
+
+package net.snowflake.client.core;
+
+import java.util.HashMap;
+import java.util.TreeSet;
+import net.snowflake.client.log.SFLogger;
+import net.snowflake.client.log.SFLoggerFactory;
+
+/**
+ * Most Recently Used and Priority based cache. A separate cache for each connection in the driver.
+ */
+public class QueryContextCache {
+  private int capacity; // Capacity of the cache
+  private HashMap<Long, QueryContextElement> idMap; // Map for id and QCC
```

Comment:
SHouldn't all those data structure be thread safe? use concurrent hash map and the like?

Reply:
may be use a Caffeine cache?

Reply:
We are synchronized two main operations i.e. Reading cache and send the QCC to GS, and writing in cache when merging elements. We think about Caffeine and conclude that it won't fit for our use case. We comment it in design doc.